### PR TITLE
Implement enough of 3d transforms spec to run the CSS FPS demo.

### PR DESCRIPTION
**Pull request by glennw**

### CODE REVIEW

**Comment by glennw** [comment]
r? @pcwalton

Once review is complete, this will need to be rebased to handle the rust-geom -> euclid name changes, but this will be very minor.

File: src/rendergl.rs [hunk 1]
```diff
@@ -9,26 +9,66 @@
...
+            x: x,
+            y: y,
+            u: u,
+            v: v,
+        }
+    }
+}
+
+static ORTHO_NEAR_PLANE: f32 = -1000000.0;
+static ORTHO_FAR_PLANE: f32 = 1000000.0;
```
Comment (pcwalton) [inline src/rendergl.rs]: These aren't from any kind of spec, I assume?

Also these should probably be `const`, not `static` (saves a load).

File: src/rendergl.rs [hunk 2]
```diff
@@ -296,6 +327,78 @@ impl SolidColorProgram {
...
+pub struct RenderContext3D<T>{
+    layers: Vec<RenderLayer<T>>,
+    child_contexts: Vec<RenderContext3D<T>>,
+    clip_rect: Option<Rect<f32>>,
+}
+
+impl<T> RenderContext3D<T> {
+    fn new(layer: Rc<Layer<T>>) -> RenderContext3D<T> {
+        // TODO(gw): This doesn't work for iframes that are
+        // transformed (but the previoud code with 2d transforms
```
Comment (pcwalton) [inline src/rendergl.rs]: typo: previoud -> previous

File: src/rendergl.rs [hunk 3]
```diff
@@ -330,273 +436,322 @@ impl RenderContext {
...
+        if context.layers.len() > 0 {
+            // Clear the z-buffer for each 3d render context
+            // TODO(gw): Potential optimization here if there are no
+            //           layer intersections to disable z-buffering and
+            //           avoid clear.
+            gl::clear(gl::DEPTH_BUFFER_BIT);
+
+            // TODO(gw): This is basically what FF does, which breaks badly
+            // when there are intersecting polygons. Need to split polygons
+            // to handle this case correctly (Blink uses a BSP tree).
```
Comment (pcwalton) [inline src/rendergl.rs]: I assume this also breaks if opacity < 1, since you can't Z-buffer those.

Comment (pcwalton) [inline src/rendergl.rs]: (BTW, this is fine for now, just wanted to make sure)

File: src/rendergl.rs [hunk 4]
```diff
@@ -609,17 +764,21 @@ pub fn render_scene<T>(root_layer: Rc<Layer<T>>,
     gl::viewport(v.origin.x as GLint, v.origin.y as GLint,
                  v.size.width as GLsizei, v.size.height as GLsizei);
 
+    gl::enable(gl::DEPTH_TEST);
     gl::clear_color(1.0, 1.0, 1.0, 1.0);
-    gl::clear(gl::COLOR_BUFFER_BIT);
+    gl::clear(gl::COLOR_BUFFER_BIT | gl::DEPTH_BUFFER_BIT);
+    gl::depth_func(gl::LEQUAL);
```
Comment (pcwalton) [inline src/rendergl.rs]: Probably worth writing a comment explaining why you need this

File: src/util.rs [hunk 5]
```diff
@@ -33,3 +41,121 @@ pub fn convert_rgb32_to_rgb24(buffer: &[u8]) -> Vec<u8> {
...
+    }
+
+    if out_vertices.len() < 4 {
+        return None
+    }
+
+    Some(out_vertices)
+}
+
+pub fn project_rect_to_screen(rect: &Rect<f32>,
```
Comment (pcwalton) [inline src/util.rs]: This is only used for clip testing, right?

File: src/rendergl.rs (hunk 1 above)
Comment (glennw) [inline src/rendergl.rs]: Correct, updated to use static

File: src/rendergl.rs (hunk 3 above)
Comment (glennw) [inline src/rendergl.rs]: Yes, you're correct - this is why we need to do the polygon splitting that blink does.

File: src/util.rs (hunk 5 above)
Comment (glennw) [inline src/util.rs]: At the moment, yes. (used to calculate the 2d rect and see if it's on screen).

**Comment by glennw** [comment]
@pcwalton OK, I have rebased this over the euclid changes.

File: src/layers.rs [hunk 6]
```diff
@@ -62,6 +89,11 @@ pub struct Layer<T> {
 
     /// The opacity of this layer, from 0.0 (fully transparent) to 1.0 (fully opaque).
     pub opacity: RefCell<f32>,
+
+    /// The depth testing mode for this layer (derived from transform-style property)
+    pub establishes_3d_context: bool,
```
Comment (pcwalton) [inline src/layers.rs]: The comment here doesn't seem to make sense for a boolean. "Whether depth testing is on for this layer" is how I interpret it—maybe reword it to say that?

Comment (glennw) [inline src/layers.rs]: Updated comment and the one below it.

File: src/layers.rs [hunk 7]
```diff
@@ -136,9 +175,50 @@ impl<T> Layer<T> {
...
+                                  parent_origin: &Point2D<f32>) {
+        let mut ts = self.transform_state.borrow_mut();
+        let rect_without_scroll = self.bounds.borrow()
+                                             .to_untyped()
+                                             .translate(parent_origin);
+
+        ts.world_rect = rect_without_scroll.translate(&self.content_offset.borrow().to_untyped());
+
+        let x0 = ts.world_rect.origin.x;
+        let y0 = ts.world_rect.origin.y;
```
Comment (pcwalton) [inline src/layers.rs]: I wonder if it would be cleaner to build a "bounds transform matrix" and "undo-bounds transform matrix" instead of rebuilding these matrices from scratch multiple times with `Matrix4::identity().translate(...)` below. What do you think?

File: src/util.rs [hunk 8]
```diff
@@ -33,3 +41,121 @@ pub fn convert_rgb32_to_rgb24(buffer: &[u8]) -> Vec<u8> {
...
+    let yc = (y0 + y1) * 0.5;
+    let vc = Point4D::new(xc, yc, 0.0, 1.0);
+    let vc = transform.transform_point4d(&vc);
+
+    let vertices = [
+        Point4D::new(x0, y0, 0.0, 1.0),
+        Point4D::new(x1, y0, 0.0, 1.0),
+        Point4D::new(x0, y1, 0.0, 1.0),
+        Point4D::new(x1, y1, 0.0, 1.0)
+    ];
```
Comment (pcwalton) [inline src/util.rs]: We're definitely gonna want to SIMDify this stuff eventually (in euclid). Paging @huonw :)

File: src/util.rs [hunk 9]
```diff
@@ -33,3 +41,121 @@ pub fn convert_rgb32_to_rgb24(buffer: &[u8]) -> Vec<u8> {
...
+    if let Some(clipped_vertices) = clipped_vertices {
+        // Perform perspective division on the clip space vertices
+        // to get homogenous space vertices. Then calculate the
+        // 2d AABB for this polygon in screen space.
+
+        let mut min_vertex = Point3D::new(f32::MAX, f32::MAX, f32::MAX);
+        let mut max_vertex = Point3D::new(-f32::MAX, -f32::MAX, -f32::MAX);
+
+        for vertex_cs in &clipped_vertices {
+            assert!(vertex_cs.w > 0.0);     // This should be enforced by the clipper above
```
Comment (pcwalton) [inline src/util.rs]: Use `debug_assert!`, no need to fail in release builds here.

File: src/util.rs [hunk 10]
```diff
@@ -33,3 +41,121 @@ pub fn convert_rgb32_to_rgb24(buffer: &[u8]) -> Vec<u8> {
...
+            let z = vertex_cs.z * inv_w;
+
+            // Calculate the min/max z-depths of this layer.
+            // This is used for simple depth sorting later.
+            min_vertex.x = min_vertex.x.min(x);
+            max_vertex.x = max_vertex.x.max(x);
+            min_vertex.y = min_vertex.y.min(y);
+            max_vertex.y = max_vertex.y.max(y);
+            min_vertex.z = min_vertex.z.min(z);
+            max_vertex.z = max_vertex.z.max(z);
```
Comment (pcwalton) [inline src/util.rs]: I filed https://github.com/servo/euclid/issues/91 to eventually fix this (so we can SIMDify it later).

File: src/util.rs [hunk 11]
```diff
@@ -33,3 +41,121 @@ pub fn convert_rgb32_to_rgb24(buffer: &[u8]) -> Vec<u8> {
...
+    // TODO(gw): Potential optimization to clip against other planes.
+    let clipped_vertices = clip_polygon_to_near_plane(&vertices_clip_space);
+
+    if let Some(clipped_vertices) = clipped_vertices {
+        // Perform perspective division on the clip space vertices
+        // to get homogenous space vertices. Then calculate the
+        // 2d AABB for this polygon in screen space.
+
+        let mut min_vertex = Point3D::new(f32::MAX, f32::MAX, f32::MAX);
+        let mut max_vertex = Point3D::new(-f32::MAX, -f32::MAX, -f32::MAX);
```
Comment (pcwalton) [inline src/util.rs]: Looks like the z coordinate is unused here. Maybe use 2D points instead?

Comment (pcwalton) [inline src/util.rs]: Oh wait, I just saw the comment. We're going to use the Z later. Never mind then!

File: src/util.rs [hunk 12]
```diff
@@ -33,3 +41,121 @@ pub fn convert_rgb32_to_rgb24(buffer: &[u8]) -> Vec<u8> {
...
+
+    Some(out_vertices)
+}
+
+pub fn project_rect_to_screen(rect: &Rect<f32>,
+                              transform: &Matrix4) -> Option<ScreenRect> {
+    let mut result = None;
+
+    let x0 = rect.origin.x;
+    let x1 = rect.origin.x + rect.size.width;
```
Comment (pcwalton) [inline src/util.rs]: There is a `max_x` function for this

File: src/util.rs [hunk 13]
```diff
@@ -33,3 +41,121 @@ pub fn convert_rgb32_to_rgb24(buffer: &[u8]) -> Vec<u8> {
...
+
+    let y0 = rect.origin.y;
+    let y1 = rect.origin.y + rect.size.height;
```
Comment (pcwalton) [inline src/util.rs]: Ditto, `max_y`

File: src/util.rs [hunk 14]
```diff
@@ -33,3 +41,121 @@ pub fn convert_rgb32_to_rgb24(buffer: &[u8]) -> Vec<u8> {
     }).collect()
 }
 
+fn clip_polygon_to_near_plane(clip_space_vertices: &[Point4D<f32>; 4])
+                                  -> Option<Vec<Point4D<f32>>> {
+    let mut out_vertices = vec!();
+    let w_clipping_plane = 0.00001;
```
Comment (pcwalton) [inline src/util.rs]: Factor this out into a constant?

File: src/util.rs [hunk 15]
```diff
@@ -33,3 +41,121 @@ pub fn convert_rgb32_to_rgb24(buffer: &[u8]) -> Vec<u8> {
...
+                previous_vertex.y + int_factor * (current_vertex.y - previous_vertex.y),
+                previous_vertex.z + int_factor * (current_vertex.z - previous_vertex.z),
+                previous_vertex.w + int_factor * (current_vertex.w - previous_vertex.w),
+            );
+
+            out_vertices.push(int_point);
+        }
+
+        if current_dot > 0 {
+            out_vertices.push(current_vertex.clone());
```
Comment (pcwalton) [inline src/util.rs]: `*current_vertex` works instead of the clone, right?

File: src/util.rs [hunk 16]
```diff
@@ -33,3 +41,121 @@ pub fn convert_rgb32_to_rgb24(buffer: &[u8]) -> Vec<u8> {
     }).collect()
 }
 
+fn clip_polygon_to_near_plane(clip_space_vertices: &[Point4D<f32>; 4])
```
Comment (pcwalton) [inline src/util.rs]: This is Sutherland-Hodgman clipping, right?

File: src/util.rs [hunk 17]
```diff
@@ -33,3 +41,121 @@ pub fn convert_rgb32_to_rgb24(buffer: &[u8]) -> Vec<u8> {
...
+            out_vertices.push(current_vertex.clone());
+        }
+    }
+
+    if out_vertices.len() < 3 {
+        return None
+    }
+
+    Some(out_vertices)
+}
```
Comment (pcwalton) [inline src/util.rs]: OK, this looks equivalent to the description of Sutherland-Hodgman I'm looking at.

File: src/rendergl.rs [hunk 18]
```diff
@@ -330,273 +436,322 @@ impl RenderContext {
...
+                    return Ordering::Less;
+                } else if a.z_center > b.z_center {
+                    return Ordering::Greater;
+                } else if a.paint_order < b.paint_order {
+                    return Ordering::Less
+                } else if a.paint_order > b.paint_order {
+                    return Ordering::Greater;
+                } else {
+                    return Ordering::Equal;
+                }
```
Comment (pcwalton) [inline src/rendergl.rs]: I think you can write this as just `(a.z_center, a.paint_order).cmp(&(b.z_center, b.paint_order))`

File: src/rendergl.rs [hunk 19]
```diff
@@ -330,273 +436,322 @@ impl RenderContext {
...
+                    let clipped_tile_rect = clip_rect.map_or(Some(tile_rect), |clip_rect| {
+                        clip_rect.intersection(&tile_rect)
+                    });
+
+                    if let Some(clipped_tile_rect) = clipped_tile_rect {
+                        let tx0 = tile_rect.origin.x;
+                        let ty0 = tile_rect.origin.y;
+
+                        let cx0 = clipped_tile_rect.origin.x;
+                        let cx1 = clipped_tile_rect.origin.x + clipped_tile_rect.size.width;
```
Comment (pcwalton) [inline src/rendergl.rs]: Use `.max_x()`

File: src/rendergl.rs [hunk 20]
```diff
@@ -330,273 +436,322 @@ impl RenderContext {
...
+                        let cy0 = clipped_tile_rect.origin.y;
+                        let cy1 = clipped_tile_rect.origin.y + clipped_tile_rect.size.height;
```
Comment (pcwalton) [inline src/rendergl.rs]: Use `.max_y()`

File: src/rendergl.rs [hunk 21]
```diff
@@ -9,26 +9,66 @@
...
+    }
+}
+
+#[derive(Copy, Clone, Debug)]
+pub struct TextureVertex {
+    x: f32,
+    y: f32,
+    u: f32,
+    v: f32,
+}
```
Comment (pcwalton) [inline src/rendergl.rs]: So I kind of wish that these were structs of Point2Ds just so that we can use SIMD when euclid gains support for it (like @huonw should be doing soon). But that doesn't have to be done now, and I'm not sure exactly how we should set this up. (Two Point2Ds? Point4D? Typed unit? Something else?) Maybe add a comment?

File: src/rendergl.rs [hunk 22]
```diff
@@ -9,26 +9,66 @@
...
+use std::mem;
 use std::rc::Rc;
+use std::cmp::Ordering;
 
-static FRAGMENT_SHADER_SOURCE: &'static str = "
+#[derive(Copy, Clone, Debug)]
+pub struct ColorVertex {
+    x: f32,
+    y: f32,
+}
```
Comment (pcwalton) [inline src/rendergl.rs]: I wonder if this can be a typed unit, so we can take advantage of SIMD once euclid gains support for it. (If it's too much trouble, don't worry about it.)

**Comment by pcwalton** [comment]
This looks good overall to me. Feel free to land once the comments are addressed, unless there are more questions. Sorry for taking so long on this!

File: src/layers.rs (hunk 7 above)
Comment (glennw) [inline src/layers.rs]: It probably does make sense. I might do that in a follow up PR though if that's OK?

File: src/rendergl.rs (hunk 21 above)
Comment (glennw) [inline src/rendergl.rs]: I think in general any heavy amount of math that would benefit from SIMD would be done while they are euclid types, and only converted to vertex structs at the end. If we ever get to the point where we feel the need to optimize these vertex types, they will probably end up being compressed types (e.g. maybe f16 for position and u8 for color), so it's probably OK to leave them as is for now.

**Comment by glennw** [comment]
@pcwalton OK, I think the last commit / comments covers those points. Thanks for reviewing.

**Comment by pcwalton** [comment]
LGTM, squash and merge away!
